src/neo4j_graphrag/tool.py

```diff
@@ -0,0 +1,261 @@
+from abc import ABC
+from enum import Enum
+from typing import Any, Dict, List, Callable, Optional, Union, ClassVar
+from pydantic import BaseModel, Field, model_validator
+
+
+class ParameterType(str, Enum):
+    """Enum for parameter types supported in tool parameters."""
+
+    STRING = "string"
+    INTEGER = "integer"
+    NUMBER = "number"
+    BOOLEAN = "boolean"
+    OBJECT = "object"
+    ARRAY = "array"
+
+
+class ToolParameter(BaseModel):
+    """Base class for all tool parameters using Pydantic."""
+
+    description: str
+    required: bool = False
+    type: ClassVar[ParameterType]
+
+    def model_dump_tool(self) -> Dict[str, Any]:
+        """Convert the parameter to a dictionary format for tool usage."""
+        result = {"type": self.type, "description": self.description}
+        return result
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "ToolParameter":
+        """Create a parameter from a dictionary."""
+        param_type = data.get("type")
+        if not param_type:
+            raise ValueError("Parameter type is required")
+
+        # Find the appropriate class based on the type
+        param_classes = {
```

Comment:
I feel like we could get rid of many of these by using either a field discriminator or some union, but I won't be strict on this point, experiments can be left for a cool down period.

Reply:
Sounds good, let's look into it later